### ISSUE TITLE
Accounts

### ISSUE THREAD

**Issue by heinrich5991** [ddnet/ddnet#3411]
If anyone wants to summarize the discussion, feel free to. I just wanted to create the issue now.

**Comment by ChillerDragon** [comment]
@Zoozti :stuck_out_tongue_winking_eye:

**Comment by ChillerDragon** [comment]
I think it is nice that ddnet is so open and all ranks/data is public and so on. But when accounts and passwords are introduced this stops.

**Comment by Learath2** [comment]
> I think it is nice that ddnet is so open and all ranks/data is public and so on. But when accounts and passwords are introduced this could stop.

This doesn't make any sense, we literally even publish database dumps sans `record_saves`. The only change I foresee is that the `Name` fields in our database becomes a `UID` field, I don't see any issue in providing a `record_users` that maps from UID to Name

**Comment by ChillerDragon** [comment]
@Learath2 but it is possible that a new database is introduced holding account data which is not public anymore right?

**Comment by Learath2** [comment]
Well obviously the authentication database wouldn't be public. Same as `record_saves` isn't public. You can't expect people to share passwords even if salted and hashed :D

**Comment by Chairn** [comment]
The real problem is how you associate old points to new accounts? Or you just don't?

**Comment by Zoozti** [comment]
Accounts are BAD.
The best thing in DDNet/teeworlds is that u can just Come and start playing instead of being annoyed with login shit.
There is no single advantage from accounts except the useless cant fake people anymore

**Comment by heinrich5991** [comment]
> The best thing in DDNet/teeworlds is that u can just Come and start playing instead of being annoyed with login shit.

Depending on the accounts design, this might not even change. If you just generate a private key on first launch, you can still add each other by friend ID, but there would be no password to create or remember.

**Comment by ChillerDragon** [comment]
> > The best thing in DDNet/teeworlds is that u can just Come and start playing instead of being annoyed with login shit.
> 
> Depending on the accounts design, this might not even change. If you just generate a private key on first launch, you can still add each other by friend ID, but there would be no password to create or remember.

This reduces awareness of what the key is and where it is located. So most people probably forget about it when replacing the drive or device. I think maybe we need some goals first what problems should the accounts solve? Is it bad when I lose access to my account?

**Comment by heinrich5991** [comment]
What problem does this solve? Provide some sort of player identity, so that finishes can be grouped together, saves don't necessarily need passwords, rcon passwords can go away except for one-time passwords, know when a friend is really your friend.

Is it bad if you lose access to your account? In that system, no, not really.

**Comment by Learath2** [comment]
Account recovery can be designed. We could use a more traditional username + password scheme and store (encrypted ?) private keys on your behalf.

Or maybe a recovery email from which you can request your private key.

Or maybe we shouldn't be doing asymmetric crypto at all. Maybe a more traditional centralized account system is more viable. We could set it up so that other servers(players?) can verify identities against our authentication server without ever learning the actual credentials.

I think we should lay down some requirements and design around that. What are the features we absolutely must have? What are features that we desire? What are the undesirables?

**Comment by houseme-kyle** [comment]
> Accounts are BAD.
> The best thing in DDNet/teeworlds is that u can just Come and start playing instead of being annoyed with login shit.
> There is no single advantage from accounts except the useless cant fake people anymore

There have been a few issues that are limited in its implementation without accounts. Heinrich mentioned a version of it that will allow for this. 

> The real problem is how you associate old points to new accounts? Or you just don't?
There is a way, but can't share it :) People interested can contact me and I'll go into the details. 

> If anyone wants to summarize the discussion, feel free to. I just wanted to create the issue now.
I'll kick off the implementation discussion below

**Comment by Learath2** [comment]
Before an implementation discussion is made I think more importantly we should have a design discussion. As I said in my last comment: We need to figure out what we want from this system.

**Comment by houseme-kyle** [comment]
> Before an implementation discussion is made I think more importantly we should have a design discussion. As I said in my last comment: We need to figure out what we want from this system.

Sorry was going to try to cover both. Certainly can do the one before the other.

**Comment by houseme-kyle** [comment]
I've tried to summarize the discussion into the three sections @Learath2 outlined. 

**Design Requirements:**
_Must haves_
- Playable without authentication 
- Authenticate client via a login mechanism
- Records and identity moved from name to a unique identifier (only necessary between client and master e.g. /w stays same)
- Support account recovery 
- Preserve old points so progress is not lost

_Desirables_
- Avoid storing confidential information

_Undesirables_
- Force accounts onto users

Please comment/add below and I'll update the list.

**Comment by Learath2** [comment]
For me "Preserve old points" is only important in the sense that the change is politically unacceptable otherwise. What I had in mind for making that palatable is to "archive" the old ladder as "legacy points" and motivating players into using accounts

**Comment by Zoozti** [comment]
> For me "Preserve old points" is only important in the sense that the change is politically unacceptable otherwise. What I had in mind for making that palatable is to "archive" the old ladder as "legacy points" and motivating players into using accounts

Bruh just stop it
No one wants ur account trash

**Comment by Learath2** [comment]
> > For me "Preserve old points" is only important in the sense that the change is politically unacceptable otherwise. What I had in mind for making that palatable is to "archive" the old ladder as "legacy points" and motivating players into using accounts
> 
> Bruh just stop it
> No one wants ur account trash

Please refrain from commenting if you are unable to keep civil or have nothing to contribute to the discussion except "I don't like it". This is the one and only warning you'll receive from me before I block you on this repository.

**Comment by Zoozti** [comment]
Im not saying i dont like it, im saying you are once again developing features against the Communitys wanting.
Why would you waste your time with that?

**Comment by Learath2** [comment]
As I've pointed it out countless times you and your 3 friends maketh not the entire community. You gave a single argument to which @heinrich5991 replied. There are many benefits to having accounts ranging from community features like whispering across servers and clans to moderation features.

**Comment by houseme-kyle** [comment]
> Im not saying i dont like it, im saying you are once again developing features against the Communitys wanting.
> Why would you waste your time with that?

It isn't constructive to give blanket statements such as "accounts are bad". Please give the reasons as to why you feel this way. 

E.g. I am against accounts because:
- I don't want to login to use the client
- I don't want to lose my points

This way others can understand the reasons behind it and can try to address your concerns.

**Comment by edg-l** [comment]
> > > The best thing in DDNet/teeworlds is that u can just Come and start playing instead of being annoyed with login shit.
> > 
> > 
> > Depending on the accounts design, this might not even change. If you just generate a private key on first launch, you can still add each other by friend ID, but there would be no password to create or remember.
> 
> This reduces awareness of what the key is and where it is located. So most people probably forget about it when replacing the drive or device. I think maybe we need some goals first what problems should the accounts solve? Is it bad when I lose access to my account?

> "Is it bad when I lose access to my account?" 

I mean, it's bad always, when would this be a good thing?

> "This reduces awareness of what the key is and where it is located."

If the key is stored in the same folder where your config is stored, I don't see how you can be "less aware".

And if you do backups of your config, adding your key should be a trivial thing.

**Comment by edg-l** [comment]
By the way, even with accounts, in the initial stages of deployment you would still not know who is who exactly, and it would be nice to find a cryptographicway to verify someone is someone, something like signing someone's key like PGP does maybe can be implemented too?

https://www.gnupg.org/gph/en/manual/x547.html

**Comment by Learath2** [comment]
Well the point of accounts would be for the server to authenticate that the person knows whatever credentials we decide on. Now if we want this to be decentralized as in a client doesn't have to rely on the server to authenticate another client, there are some issues that come into play, like [Zooko's Triangle](https://en.wikipedia.org/wiki/Zooko%27s_triangle).

I'm sure we could come up with a semi-centralized solution to this. Which is mainly why I wanted to discuss the design first.

**Comment by houseme-kyle** [comment]
I have updated the list to have "decentralized" as a desirable feature. My feeling is that if it doesn't impact the complexity of the solution too much and the usability of it, then it is worth considering. 

@Learath2 I'm not sure if the format I provided and the points are along the lines of what you had in mind.

**Comment by houseme-kyle** [comment]
Alright, there seems to have been some consensus on discord. The solution was to authenticate via emails to avoid handling passwords amongst other benefits.

**User Flow**
1. /login <email> in server || alternative is to have a login UI button
2. /otp <email> <code> in server || alternative is to have a dialog post clicking login
3. Success! 

// There is a solution without OTPs. I am unsure though of how that will be communicated back to the client from the email.

**Implementation suggestion**
1. Client sends the email address to the master server
2. Master server sends an email with a pin to the address
* Pin is a unique identifier generated off the email similar to a hash
3. Client sends the pin, email address, name, country code to the master server
4. Master server checks if the pin matches the email address. If so, sends back access_key and upserts account info

//Master server doesn't have to be the same as the current
Thoughts on how the database can be structure:
https://dbdiagram.io/d/604cb0bdfcdcb6230b23fa0e (feel free to make comments in-line - there is revision protection)

**Comment by heinrich5991** [comment]
This database structure looks a lot more complicated than what we have today. What is the motivation for this?

**Comment by houseme-kyle** [comment]
I decided to whip up a Proof of Concept application for the above. I'm fully aware that it is likely the project will need to be re-written as it is in an unused language. 

My objective was to create a secure account system while still adhering to the principle of an open database. This is achieved through a static secret salt and generated salt per pin. 

1. User requests login 
2. Entry into PinTable made with:
- Lookup: Hash of email address and secret salt
- Readable Code: Deterministic Guid created from the email address, salt, and a different secret salt
- Valid Timestamp: 12 hours
3. User supplies readable code and email address - client will send through name and country code as well
- Lookup: hash of email address and secret salt. If still within valid period, fetches readable code
- Compares associated readable code with the supplied code
- If valid gets or creates the user with a new access key
4. Future account requests to be made with the access key. 

I'm no security expert so welcome any feedback. Thought it would be useful to get something functional out so we can start poking holes at the implementation. 

**Code**
https://github.com/houseme-kyle/DDNet-Api
_My code isn't production-ready so please no comments about that_ 

**Database**
URL: ddnet-temp.mysql.database.azure.com
User: readonlyuser@ddnet-temp
Password: password

**App**
https://ddnetclientapi20210315002458.azurewebsites.net/swagger/index.html

**Test inbox**
_Email provider requests a verified email address in sandbox_
_This step can be bypassed as the code is returned for now from the call_
https://www.guerrillamail.com/inbox
pincodecatcher@sharklasers.com

**Comment by houseme-kyle** [comment]
> This database structure looks a lot more complicated than what we have today. What is the motivation for this?
I find it useful to conceptualize the potential future requirements as a safety check. 

Breaking it down a little:
_Account related tables_
- Users: Added for accounts
- PlayerRoles: Needed when wanting to grant permissions based on role e.g. tester, mapper, admin etc.
- Clans: Needed when wanting to support clans

_Existing tables - not exactly part of accounts but a chance to re-assess_
- RaceCheckpoints: I noticed the majority of maps had these columns unused and when they did it was variable. Felt it could benefit from being its own table for that flexibility 
- DifficultyTiers: There is a coupling between difficulty levels, tiers (i.e. stars) and points. Unless I am mistaken, you wouldn't have a map with the same level, tier but different points. This allows for a singular location to maintain that association 
- Server: This logic is housed informally somewhere, so useful to bring it in here where the data is structured and queryable.

The "existing table" changes are minor iterations and I suppose aren't important for the initial account discussion.

**Comment by heinrich5991** [comment]
Design as @Learath2 and I talked about:

Each client generates an asymmetric key and uses it to authenticate against each game server. This is transparent to the user.

The user can "log in" to an account by giving an email address while authenticating with its private key. The account server will store a constant, never changing mapping of public key hash → account ID. This mapping is public and can be queried via an HTTPS endpoint so third party servers can also use it. The fact that the mapping is constant can be used to aggressively cache the mapping on all servers, lowering the impact if the authentication server is down.

A user might play from multiple clients – the private keys aren't shared by default, instead the user logs in from multiple private keys.

There are two methods I can think of to support logout. Either the client generates a new key (might have a bit of a bad UX because we then don't trust the client anymore), or we include an increasing login index, and use hash(public key + login index) as login identifier instead of hash(public key), meaning that logout would simply be incrementing the login index.

In order to securely support login, the client-server connection must be encrypted and authenticated, otherwise MITM and proxying issues immediately appear. The login protocol should bind to the cryptographic session so that the login sequence cannot be forwarded to other servers.

**Comment by Scrumplex** [comment]
This system could also be used to combat spam.
For example: Authenticated players with x amount of points are not affected by network-wide mutes for new players.

It should also be "hard" to create many accounts with something like email verification or some kind of CAPTCHAs (undesirable imo).

**Comment by heinrich5991** [comment]
@edg-l wrote:

> I was bored so I did this, not sure how useful it is:
>
> This is a really simple backend for the auth server using ed25519 keys because they are short. I made this reading this (https://github.com/ddnet/ddnet/discussions/4033#discussioncomment-1184498)
>
> https://github.com/edg-l/ddnet-authservice
>
> Endpoints
>
> | METHOD | Endpoint | Data | What it does |
> | --- | --- | --- | -- |
> | GET | `/version` | Auth server version, e.g 1.0.0 | Tell the backend version |
> | POST | `/account/register` | A Json object "RegisterPayload" (check below for how its defined) | Register a new account |
> | POST | `/account/mapping` | A json object "AccoundIdRequest" (check below for how its defined) | Given a public key, get its mapped uuid |
> | POST | `/account/verify` | A json object "VerifyUserPayload" (check below for how its defined) | Verify a public key is valid and return the account id |
>
> ### RegisterPayload
> ```json
> {
>     "public_key": "base64 encoded public key byte data",
>     "email": "email@email.com",
>     "email_signature": "base64 encoded signature of the email using the private key for the given public key",
> }
> ```
>
> ### AccoundIdRequest
> ```json
> {
>     "public_key": "base64 encoded public key byte data",
> }
> ```
>
> ### VerifyUserPayload
> ```json
> {
>     "public_key": "base64 encoded public key byte data",
>     "message": "something here",
>     "message_signature": "base64 encoded signature of the message using the private key for the given public key",
> }
> ```

@Astramast wrote:
> Hey we wanted to do an account system right ?
> for ppl to not be able to do $hit like racial slurs in chat and also to have friendlists account-bounded
> So KoG actually have an account system
> We could use a copy of their system 0w0
> Also would be possible to chat-restrict guests
> Like u cant talk if ur not logged in
> Or create an option in chat settings "See guest messages"

@Astramast wrote:
> Personnally i think that accounts should only serve to certificate player (friendlist and chat security). For the finishes, we could keep the actual system and add over it the UID system (ex : if i search Cor in ddnet.tw/players i will find every finish made by a tee that has the name Cor and then u would be able to search for [UIDCor] and u will have only the finishes made by the account Cor (wich are all in the 'Name' search) a bit like a filter)

@Astramast wrote:
> With this system Newplayers could totally play normally and when they create their account it will copy all the finishes that are in the 'Name' into the UIDName and after only the finishes made by loggedin account (that should be automatically done at connection like in KoG) will be registered in UIDName

**Comment by PotatoCreator** [comment]
Hello, there! I like the idea about accounts. I myself don't know how coding and stuff work. I am here to tell my opinion about this. I see it as good improvement since it would fix many problems. I will share what is my opinion about this feature. The system may not only have an option to auth, but much more.

This idea could be working either with steam account/or ddnet forum account for easier password reset.
Register/Login could be on first screen of ddnet client or on server itself.
On server there would be a bind available to join in 1 click after successful registration
or if in main menu, Button "keep me logged in" could exist

Ideas if this gets real:

- Free possibility to check your saves on certain map or in general.
- User statistics (optional) (Timespent, Fav server, etc) 
- Contact info (If user wants to allow it, Discord, Steam, etc)
- User rules breaking/ban history
- /report (user) for quicker access to report a rule breaker.

Pros:

- User faking would disappear.
- Moderation simplifies.
- Tracking violations possibility 

Cons:

- Basically spending 2 minutes of your life to create an account :/ 
- Forgets his steam/ddnetforum/discord passwords

**Comment by def-** [comment]
My approach for accounts would be quite simple:

1. Central account server to register on and log in to administer your account
2. Account server returns private key token to player, stores public key token
3. When client joins a game server, game server asks account server for the currently valid public key tokens
4. Game server sends challenge to client, client uses its private key token to solve it. Thus evil server hosters can't steal tokens and impersonate players.

Allow every username to be registered an unlimited number of times, assign a random integer to the username (similar to Discord). Players add friends using the name+id. Nicknames can be used ingame, but actual name+id are also shown. Thus we have no effort to make sure players are claiming names of known players already.

Don't move points from old accountless world. Keep them around as non-account players and players can keep using them forever. Alternatively enforce accounts and stop supporting playing without accounts after a few weeks of transitory period. I'm not sure which I prefer.

Start a season 2 of DDNet instead where account players start at 0 points, but can collect new achievements, only available for logged in players. Something like the Duolingo achievements:
![Screenshot 2022-11-13 at 14 35 06](https://user-images.githubusercontent.com/2335377/201524493-34fd02bd-1034-434c-8599-5a3b9262ec88.png)

This should motivate players to start using accounts. When achievements are gained they should be shown in client with some small animation, and on website.

**Comment by heinrich5991** [comment]
> 4\. Game server sends challenge to client, client uses its private key token to solve it. Thus evil server hosters can't steal tokens and impersonate players.

Evil server hosters can still impersonate players: If someone joins the evil server, it can join a real server and forward the challenge from the real server to the player. The player solves it and now the evil server impersonates the player.

**Comment by heinrich5991** [comment]
> 2\. Account server returns private key token to player, stores public key token
> 3\. When client joins a game server, game server asks account server for the currently valid public key tokens

It would probably make more sense to generate the private key on the client side. I think it'd make more sense for the game server to ask for the account related to the public key the client just proved.

**Comment by Kaffeine** [comment]
See also: https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol (implements the idea of "Zero-knowledge password proof" authorization).

**Comment by Scrumplex** [comment]
We could take inspiration from OAuth2 and let the auth server do the actual challenge. All the server needs would just be a :+1: from the auth server for that particular player. To avoid MITM and malicious proxies, the auth server can also tell the server the actual client IP. So if connecting address and authenticating address do not match, we can just reject the join request.

**Comment by Vy0x2** [comment]
Would be cool if the server is able to know when the account got registered.

**Comment by def-** [comment]
Ok, the problems with malevolent servers impersonating other players is annoying.

**Comment by l-ouis** [comment]
> Start a season 2 of DDNet instead where account players start at 0 points, but can collect new achievements, only available for logged in players.

I'd be very careful with how to go about season 2 - don't think it will be very popular unless it adds/changes a substantial amount of the game.

**Comment by def-** [comment]
Yeah, that's what I'm also worried about. But if we want to keep non-account ranks, then we'd have to manually merge them with accounts, which will be a lot of effort and affect many thousands.

**Comment by def-** [comment]
Next steps, some can be done in parallel:

1. Check if the community likes this idea, done: https://discord.com/channels/252358080522747904/420565311863914496/1041489111174877244 (blocking)
2. Figure out how to safely get authentication data from other players to your client. We could only trust the http master server for this, which we also run, and not the game server. Registration server could tell the http master who is logged in where.
3. Implement a backend to register accounts from website/client
4. Implement a frontend to administer your account from website/client
5. Think about what achievements we want to have and make them show up on website/client. Some should be stackable so you can improve on them:
  a) Number of days with finishes every day: 7 days / 14 days / ...
  b) Long finish: 4 hours / 8 hours / ...
  c) Reach points: 1000 / 2000 / ...
  d) Team rank with same friend on number of maps: 10 / 20 / ...
  e) Perfect time finish with :00.00
  f) Team finish with large number of players: 16 / 32 / 64
  g) Top ranks: 1 / 2 / 4 / 8 / ...
  h) All maps finished on server: Novice / Moderate / ...
  i) Account age and still active: 1 year / 2 years / ...

**Comment by Anime-pdf** [comment]
I think its step in right direction, ddnet should progress and eventually deny old principles, hope next step will be stopping integration with vanilla and 07 with upgrading network protocol and adding more customisation to client from server side

**Comment by l-ouis** [comment]
> Figure out how to make registration for Steam users "one-click". Can we get their Steam ID and connect them to Steam automatically to sign up? Then no email and password required.
fwiw, a lot of games have 2 options on new player join: auth thru steam and create a "[company name here]" account. I think having both options would be good for those on steam that don't want their DDNet account to be tied to steam

> Are there any other major things we should change now that we have the chance of a "new season"? Things which could not be changed normally because of backwards compatibility?

If you are gonna make a new season, go all out
The way I see it now, if you keep all maps and change nothing:
- people will be very excited at first to grind for some achievements
- people realize they will have to play the same maps they already played for the achievements, burn out, and then it won't be fun anymore

I think: don't call it a season 2, call it a ddnet 2 (or smth similar) and then:

1. remove janky game mechanics (mainly tiles)
2. replace some janky game mechanics with better coded mechanics (looking at speeders and the such)
3. add some new cool weapons and tiles
4. improve the map rating and ranking system
5. start the map pool from scratch

among other changes

**Comment by heinrich5991** [comment]
> See also: https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol (implements the idea of "Zero-knowledge password proof" authorization).

This requires a secure connection, otherwise it has the same problem with forwarded auth challenges.

> I personally think it would be nice if ddnet only provides a central account server so managing the account is good for non-IT ppl.

I really don't understand why you think that "managing an account" is hard when the backend uses public-private key cryptography. Just treat the public-private key stuff similar to an authentication cookie. You wouldn't call logging into Wikipedia hard because you get some opaque token.

> 2\. Figure out how to safely get authentication data from other players to your client. We could only trust the http master server for this, which we also run, and not the game server. Registration server could tell the http master who is logged in where.

A secure server connection (doesn't need to be authenticated) would also make that possible. I guess that's too far future?

**Comment by edg-l** [comment]
Going for a ddnet 2.0 and breaking compat to cleanup the protocol and stuff sounds very appealing to me not gonna lie.

The thing is, if we try to couple the accounts change with more changes its possible we will never deliver anything, although if properly planned maybe.

Personally i would love a complete rewrite of the mess that is the website code, which is quite unmaintenable for anyone that is not deen, the scripts and static part being in 2 separate repos etc. We could move on from static generation to a more dynamic website, since we have so much data, and so much potential, and now we could even the in-game accounts to the website and provide more features, etc.

**Comment by Zoozti** [comment]
Accounts will kill the Game, make a Lot players ragequit and also will Stop new people from starting the Game because Nobody wants to register for a 2008 2d Game. Keep it what an Open source Game should be. Free and easy to handle.

And really deleting all finishes?xd 
Like removing years of Players Lifetime for nothing

**Comment by Zoozti** [comment]
Ur putting the Vote of some random new Steam Players over the Work and time other Players spent in 10 years

**Comment by kyle-bradley** [comment]
You don't need to force registration. When I implemented accounts for the
SA community it was optional and in the end had almost full uptake from all
regular players.

I would also suggest you still recognize the achievements of the old-guard,
while future proofing the game with the new system. Few examples:

   - Have awards for season 1's players
   - Give recognition on the new profiles for the top X players
   - Show "All-time" records that include pre-account times

 The above are a few ideas which aren't perfect but I believe it is
possible to achieve a balance of both.

**Comment by def-** [comment]
@l-ouis @Anime-pdf I think your suggestions go too far. If we make a DDNet 2, then we have to keep DDNet 1 around, and it will be difficult to get players to move to DDNet 2. Also, it's too much work, at least for me. Accounts + Season 2 is probably the most we can realistically manage in the next few months.

**Comment by Anime-pdf** [comment]
> @l-ouis @Anime-pdf I think your suggestions go too far. If we make a DDNet 2, then we have to keep DDNet 1 around, and it will be difficult to get players to move to DDNet 2. Also, it's too much work, at least for me. Accounts + Season 2 is probably the most we can realistically manage in the next few months.

Probably, but i support louis's map pool idea, if you keep same maps it will look like you devalue the work of the players, maybe keep all maps that released in 2022 or smth like that, but noone would like to grind same maps again

**Comment by def-** [comment]
If we do that then someone else will just host all maps and they are the new DDNet, not us.

**Comment by Anime-pdf** [comment]
Thats a dead end situation

**Comment by Zoozti** [comment]
Or Just keep ddnet like it is. I wouldnt like Losing my ranks i spent 9 years in

**Comment by Zoozti** [comment]
> > Or Just keep ddnet like it is. I wouldnt like Losing my ranks i spent 9 years in
> 
> There are 804,394 players with points, if i understood `ddnet.org/stats/` right, noone willing to assign account to each of them, and if we include trolls like you, who would likely be faking someone's accounts, its nearly impossible. Ddnet must progress and accounts are what game really needs, we are discussing it rn because we dont know the best solution yet

I dont see a single reason why the Game would need accounts.
Teeworlds exists for 15 years and it works perfectly fine without Trolls like U forcing people to register for shitty Accounts and losing all their progress

**Comment by Anime-pdf** [comment]
> > > Or Just keep ddnet like it is. I wouldnt like Losing my ranks i spent 9 years in
> > 
> > There are 804,394 players with points, if i understood `ddnet.org/stats/` right, noone willing to assign account to each of them, and if we include trolls like you, who would likely be faking someone's accounts, its nearly impossible. Ddnet must progress and accounts are what game really needs, we are discussing it rn because we dont know the best solution yet
> 
> I dont see a single reason why the Game would need accounts.
> Teeworlds exists for 15 years and it works perfectly fine without Trolls like U forcing people to register for shitty Accounts and losing all their progress
> 

Ddnet is successor of teeworlds, it should be same 15y.o game, accounts will give more capabilities for mode makers, event organisers, it will be easier to integrate in any app and will help with reports, fake finishes, it will give clan system a second breath after all(maybe)

**Comment by Scrumplex** [comment]
> I dont see a single reason why the Game would need accounts.

- "Verified" finishes (i.e. you don't need to worry about other people finishing maps with your name)
- More effective spam protections
- Harder impersonations of users

I am pretty sure that your old points won't be gone forever. They will just be frozen in time.

**Comment by edg-l** [comment]
> This is a dead end situation, resetting ranks with same maps = loosing players. New maps instead of old = loosing players.

Realistically, thats false, ddnet keeps growing and players that may leave due to this are few or none, the influx of new players is higher.

Also konsti, you are no one to dictate how ddnet should be ran, as you said, this is a free open source game, you are free to host your servers, make your own network, download the current ddnet database, and run it.

Accounts have obvious benefits:
- it gives players more authorship over their achievements, you can keep track of their records, history, etc. 
- Makes extreme trolls easier to handle, for example, there could be a lockdown chat mode in servers where players with accounts not older than x days or minutes can't talk.
- You can have badges that promote involvement in the community: map tester, developer, wiki editor, moderator, translator, tournament winner, donator, etc
- Other players can't finish under your account unless you explicitly go out of your way to handle them your credentials, right now anyone can finish for you even if you don't want it.
- Maps can be bound to author accounts
- Skins can be bound to author accounts

**Comment by ardadem** [comment]
> I dont see a single reason why the Game would need accounts. Teeworlds exists for 15 years and it works perfectly fine without Trolls like U forcing people to register for shitty Accounts and losing all their progress

It is not working well and it is hard to moderate. Trolls keep joining the game with new nicknames and using proxy, fun voting or start blocking people, spamming on the chat, using your nickname to finish maps etc. Yes, it was easy to moderate earlier days with a small community, but it is not now. Having an account mechanism is the best solution here.

Also instead of achievements, skins as a reward would be more motivating.

**Comment by Zoozti** [comment]
Implementing Accounts is the opposite of a free open source game.
Y'all pretending Like your the President of the united States and everyone is trying to fake you lmao.
There is nothing actual usefull in accounts which could be worth removing 10 years of Players History and time spent into finishing the hardest maps

**Comment by Zoozti** [comment]
> 

lol there is like 1000 skins u can freely choose and Most people dont Change them in years. Why would anyone Go for a shitty skin.

And yeah, some Trolls spamming in the chat which U can easily mute by one click is for sure a good reason for accounts

**Comment by Anime-pdf** [comment]
> > 
> 
> lol there is like 1000 skins u can freely choose and Most people dont Change them in years. Why would anyone Go for a shitty skin.
> 
> And yeah, some Trolls spamming in the chat which U can easily mute by one click is for sure a good reason for accounts

Yeah, 100 people muting one guy locally is much better than muting this guy for everyone

**Comment by Learath2** [comment]
As a master spammer yourself I'm sure you are well aware that you can dodge mutes fairly easily. Which is the real reason you are here arguing right now. To keep your ability to be a nuisance.

**Comment by Scrumplex** [comment]
> Implementing Accounts is the opposite of a free open source game.

Apart from principle do you have any actual reason against accounts?

Privacy?
Doesn't matter as DDNet can already track all your activity by the username you choose and the ip address you connect with. Or maybe even with your timeout token. An accounts-system will not change this.

FOSS?
I am sure that all components will remain free and open source. If you don't trust DDNet's servers, you can just go and build your own network with the same tools.

**Comment by l-ouis** [comment]
What do you all think of the legacy map pool idea? I think it's important to devalue the worth/importance of current maps if a season 2 would be released.
It makes it so that players who grinded points/maps don't feel the need to "restart" their progress, since old maps will be in a legacy pool where they do not matter as much
It also makes new maps and rankings a lot more competitive. Previously, many many DDMax or Oldschool maps would only be played for skips or farming rankpoints since they'd have a lot less speedrunners. With a fresh map pool, DDrace would be a lot more competitive

**Comment by Zoozti** [comment]
> As a master spammer yourself I'm sure you are well aware that you can dodge mutes fairly easily. Which is the real reason you are here arguing right now. To keep your ability to be a nuisance.

Talking Nonsense as usual. Ur not even playing teeworlds at all but always think u know whats the best Thing for Community.
Changing teeworlds physics, adding nobo hammer, implement accounts Just to Count a View.

Im against accounts because its destroying the Soul of the game and also there is No way to verify the current Players so the only solution is to take thousands of hours of peoples Playtime away for your stubborn Views.
Also the discord vote is super rigged because Chinese Players make the hugest amount of players but they cant even enter discord

**Comment by Brokecdx** [comment]
Accounts open up countless opportunities for the game to develop and increase its player base. 
Achievements, rewards, special bound stuff etc. 

However I am playing this since a very very long time and I have to admit, my recent grind to be lost would be hard to bear. All the rank1's we grind for hours just to get super sick time would be less important. Any effort would be meaningless in a way. 

Obviously there would be new challenges however it also means the newer maps released post 2020 that we grind rank1 at would also be lost. Especially the insane category grind would have 0 impact and 0 rewards for us. I personally like the idea but there has to be somehow rewards for already done rank1s perhaps on legacy then, something special, that you can only achieve with legacy maps then.

I think a lot of people would mind the change or rather the "hard reset" less if they could get something back for the time being spent the past 9 years. 

So I think to satisfy both parties we should together think about a solution that allows to hard reset a season2 and start fresh but also reward the time spent prior Season2, the rank1s made, the miracles that players have done, tournaments, points, teamranks etc.

Nobody should be left behind that has done something meaningful on record. (With necessary proof of course)



Conclusion: Accounts yes, but with the best solution possible to both satisfy new gen and old gen :)

**Comment by l-ouis** [comment]
> So I think to satisfy both parties we should together think about a solution that allows to hard reset a season2 and start fresh but also reward the time spent prior Season2, the rank1s made, the miracles that players have done, tournaments, points, teamranks etc.

I think the only solution, since linking current players to old names is not feasible, would to be somehow freeze the old ranks and stats somehow (like some hall of fame? idk)

**Comment by def-** [comment]
My idea would be to have a separate page for Season 1 and Season 2. You can link your season 2 and season 1 profiles together if you want, similar to the /compare/ feature: https://ddnet.org/compare/deen/deen-8322-/ But we don't validate that in any way, it's not officially connected to your old account. But if you just want to see your total points of season 1 + season 2, there you go.

**Comment by heinrich5991** [comment]
> Talking Nonsense as usual. Ur not even playing teeworlds at all but always think u know whats the best Thing for Community. Changing teeworlds physics, adding nobo hammer, implement accounts Just to Count a View.

Stop the personal attacks.

> Im against accounts because its destroying the Soul of the game and also there is No way to verify the current Players so the only solution is to take thousands of hours of peoples Playtime away for your stubborn Views.

I agree that taking away our current ranks sounds bad. I don't know if that's something DDNet will recover from, and I think we should be careful not to kill the game.

> Also the discord vote is super rigged because Chinese Players make the hugest amount of players but they cant even enter discord

Do you have a suggestion that is better to gauge community interest in accounts?

**Comment by bubliman** [comment]
@kyle-bradley 
> People interested can contact me and I'll go into the details.
I have been wondering about this, would like to talk. Could u give me a contact? or hit me up on discord bubliman#2668

**Comment by Desour** [comment]
Hi, I'd like to give some input as a player.

* I do not want to have to `/login` each time I join a server.
* I don't want to *have to* manually register somewhere in order to get verified points / ranks.

I'd therefore advocate for the design by heinrich5991 and Learath2, where the client automatically generates a key pair.



However, for security reasons (i.e. MitM attacks you already mentioned), as long as ddnet uses unencrypted network, I'd like to propose there to be a web-based account system on ddnet.org instead of an in-game login with email. One could then upload one's public keys there, and do other stuff, like withdrawing old keys. To verify that you actually own the private key, a key upload can then be confirmed in-game via chatcommand where you enter a validation code you got from the website.

(Also, I do not see a reason for a logout feature.)

----

> > 4. Game server sends challenge to client, client uses its private key token to solve it. Thus evil server hosters can't steal tokens and impersonate players.
> 
> Evil server hosters can still impersonate players: If someone joins the evil server, it can join a real server and forward the challenge from the real server to the player. The player solves it and now the evil server impersonates the player.


For authing with your public key, you could include the connection details (i.e. IP and port of client and server) to the auth challenge. This way, the client wouldn't authenticate a session for a server that it's not directly connected to (assuming there's no IP spoofing).

**Comment by Scrumplex** [comment]
For the average player's sake, I am strongly against an asymmetric key system for authentication, if it's not automated and seamless.
Players won't want to deal with this. People are going to lose their keys because no one has proper backups. This is going to generate an unnecessary amount of load for DDNet staff.

Just keep it simple. Create a simple centralized account system. Do all the session stuff over HTTPS with that account service. Add SSO using Steam for convenience.

About transferring finishes: Perhaps it's possible to build an automated system that can match player's demos with tee historian demos for verification. That way, we could require a few demos and attach a legacy username to an account in the new system.

The account system really just needs to keep track of email addresses, usernames, passwords and some extra info for OAuth using Steam.

I don't buy into any of the privacy concerns, as people could just use a throwaway email to register initially.

**Comment by heinrich5991** [comment]
> For the average player's sake, I am strongly against an asymmetric key system for authentication, if it's not automated and seamless.
> Players won't want to deal with this. People are going to lose their keys because no one has proper backups. This is going to generate an unnecessary amount of load for DDNet staff.

This is why the proposal above is not about manual asymmetric key system.

**Comment by Zoozti** [comment]
Just finally and completely drop this years old accounts idea. It didnt even work for reasons when we were a small community, it will not work now and not in future and theres no need for it. 
People love to simply play, whos going to convince all the chinese players from accounts? It takes the simplicity out of the game and you cant want that

**Comment by ewancg** [comment]
I am of the opinion accounts should be implemented. It would put a stop to many forms of denial-of-service attacks (which are becoming increasingly popular, at least where I play), force players to stick to the rules, and allow better integration with other services. It's hard to hold players accountable when ban evasion is as simple as resetting the router.

If implemented, I think accounts should be mandatory for playing on official DDNet servers. Fake-naming is bad, even when not being used as a device for ban evasion. People who fake for the fun of it think they're much funnier or more mysterious than they actually are. It creates confusion among players & generally has no legitimate reason to be allowed besides the principle of player freedom.

**Comment by Zoozti** [comment]
> I am of the opinion accounts should be implemented. It would put a stop to many forms of denial-of-service attacks (which are becoming increasingly popular, at least where I play), force players to stick to the rules, and allow better integration with other services. It's hard to hold players accountable when ban evasion is as simple as resetting the router.
> 
> If implemented, I think accounts should be mandatory for playing on official DDNet servers. Fake-naming is bad, even when not being used as a device for ban evasion. People who fake for the fun of it think they're much funnier or more mysterious than they actually are. It creates confusion among players & generally has no legitimate reason to be allowed besides the principle of player freedom.


Yeah, go and make accounts mandatory and see your favorite game dying because zero new players will come as the game is already complicated enough.

**Comment by ewancg** [comment]
The majority of game downloads (and as such, new users) go through Steam nowadays, I'd be surprised if this account system didn't integrate into Steam so registration is either simple or requiring no extra user action at all. As for existing players -- I doubt that anyone who's played this game more than a few times will just up and quit because they have to create an account. If done right, it will take a few minutes at most. It's not as big of a deal as you're making it out to be.

**Comment by heinrich5991** [comment]
> It would put a stop to many forms of denial-of-service attacks (which are becoming increasingly popular, at least where I play),

I don't think accounts are something that can help against DoS attacks. Can you explain how that would work? Either there's a server bug to be fixed (independent of accounts), or it's a volumetric DoS attack that you can't do anything against anyway.

> Fake-naming is bad, even when not being used as a device for ban evasion. People who fake for the fun of it think they're much funnier or more mysterious than they actually are. It creates confusion among players & generally has no legitimate reason to be allowed besides the principle of player freedom.

I don't see how the reasons you give why "fake-naming" is bad are valid. I don't think you can use "fake-naming" as ban-evasion (our bans are IP-based). Why is "fake-naming" a problem in DDNet but not in Steam/Discord? I don't think we need to disallow people choosing the name "heinrich5991" or "deen" after introducing accounts. Can you elaborate on these benefits?

**Comment by ewancg** [comment]
> I don't think accounts are something that can help against DoS attacks. Can you explain how that would work?
If a new connection doesn't pass the auth. step, which would be based on this account system, add it to a blacklist (iptables or something). This vastly reduces the area that can be attacked, because the malicious actions are never reaching the point in the game code where it could trigger an exploit. This does not protect against volume-based attacks or attacks based on a bug or oversight in the auth. step.

> I don't see how the reasons you give why "fake-naming" is bad are valid. I don't think you can use "fake-naming" as ban-evasion (our bans are IP-based).
Technically you can re-join a server you were banned from with the same name as long as your IP is different. But you will just be banned again in a matter of minutes. Fake-naming is not a problem on either of those platforms because they both have some sort of unique identifier for the user beyond just a name. They also have systems in place to make ban evasion much harder (iirc, Discord will ban every account from a server if they are logged into on the same IP as the user who was originally issued the ban). And the benefits? you know who you're playing with!

**Comment by Kaffeine** [comment]
> If a new connection doesn't pass the auth. step, which would be based on this account system, add it to a blacklist (iptables or something).

I know it is different for different servers but at least a part of data center operators do measure/analyze the traffic before it is passed to the hosting. In other words, even if I block literally *all* traffic in iptables (I can control the machine directly via admin panel), I still get messages:
> Dear Customer,
> 
> This is an automated message to inform you that your IP address xxx.xx.xxx.xxx has been nullrouted due to DDoS attack (incoming UDP 751.79 kpkts/s)
> Nullroute will be removed automatically after <N hours> from now. No actions from your side are required.

`¯\_(ツ)_/¯`

**Comment by Jupeyy** [comment]
I'll try to explain my proposal again. I think it's close to what heinrich would what, but there is one detail where it differs completely:
How the game server identifies the user. How the private key is stored. The communication level with ddnet.org

# Terminology

game-server = a server or a group of servers that is hosted, either by ddnet or other communities. all kog servers would be one game-server in this case for simplicity
recover-id = An ID generated for the game-servers and the client (communication between both) to make sure that certain actions like recover are identifiable

# Disclaimer

I want accounts to be fully optional. Imo just playing and having fun is important for this game. I fully understand that this can also be a problem, e.g.
it __might__ be harder to moderate without forced accounts (and with forced accounts I mean actually login, not some auto generated key pair).

# Overview

I want three possible ways to to have an account:

1. You have a normal login/password on ddnet.org
2. You have some kind of external auth, that also registers to ddnet.org (e.g. through steam login)
3. You don't have any relation to ddnet.org, but instead manage your key-pair your own. (unsafe for a non experienced user)

# The implementation detail

All three methods have one thing in common:
All users have one unique ID (probably globally shared accross all servers, (easier for friend lists etc.) <- no strong opinion if it could also be per game-server).
They all use a key-pair PER game-server, account related communication ALWAYS happens over the encryption by this key-pair:
1. & 2. stores the key-pair (or more accurately, the private key) for you. If you switch computers, u simply redownload it after the login to ddnet.org. So basically a fully managed account. Imo we can also try to never store the private key to the disk at all. That would enhance security by a lot, bcs the user has a harder time to share the key with a criminal.
3. You have to make sure you don't loose the private key, it's fully manually.

# Registration

the game-server somehow signals the client, that is supports accounts:
1. & 2. The user asks ddnet.org to register to a new game-server -> ddnet.org sends some unique ID (for recovery, more later), ddnet.org generates a new key-pair for this game-server -> user sends the unique ID and communication happens encrypted over the key-pair
3. client generates a key-pair and sends an E-Mail or Nothing to the game server for recovery

# Identification

All users have a unique ID that is never ever changing. The user has no interest in faking this ID, this ID alone is not a full identification!. This ID alone is ofc not secure so:
All users are additionally identified by the public key. It's up to the game server how exactly.. Most commonly probably the hash of the public key.
Nothing less, nothing more. DDNet.org does not need to be pinged at all. There is no list of key-pairs the server needs to download (heinrichs approach IIRC)

# Recovery

The game-server additionally gets something that allows you to recover an account (as teasered):
The client also knows about this ID, to have a way to communicate about a specific action (e.g. like this recovery).
This is the only time, the game-server has to communicate with ddnet.org directly (It's basically as if ddnet.org is your E-Mail address wrapper)

1. & 2. It has some unique ID (the recovery-id). This ID is generated once by ddnet.org, when the user wants to register to a game server. The game-server uses this key and a custom recovery key generated by the game-server to ask ddnet.org to recover the account (basically like an recovery e-mail wrapper):
    for 1. ddnet.org sends an actual e-mail with yet another recover key(generated by ddnet.org this time) to make sure the request is valid. if the user indeed wants a reset, ddnet informs the game-server (with the game-server recovery-key) (or the game server simply asks ddnet again if the request was valid (the client can always say "i recovered my account, using THIS recovery key") & ddnet.org generates a new key-pair ( & maybe recovery-key ) for this game-server -> the game server knows that the identity changed and can take actions
    for 2. e.g. for steam, simply re-request a login token (if your steam account is stolen, ddnet also can't really do more here I guess). but similary to 1.
3. an e-mail is given to EVERY game-server, you registered to. The game-server can send a recovery key to this E-Mail. If no e-mail is given.. then you have to live with the loss.. Again this is the unsafe variant most ppl should not use. Just for hipsters like me ;)

# Automatic register

If someone has no account, we can still make a key-pair in the client for every game server and auto register, similary to how heinrich imagined it. I think my approach is not blocking this idea

# Summary

- A normal user creates an account on ddnet.org (normal Login/Password or steam login)
- User joins a server. Server tells the client "I have accounts, do you want to register?". (Maybe user has to press yes. maybe it happens automatically <- no strong opinion on that). Client tells ddnet.org -> ddnet.org makes key-pair & recovery key. An account on a game server is basically a variant:
    * ddnet.org account
    * custom account
    * no account
- User and server communicate over key-pair encryption.
- public key = the account ID
- user get's hacked -> hacker has key-pair (but cannot change this key-pair without the e-mail) -> user can recover his account. Asks game-server to recover (recovery-id) -> game-server asks ddnet.org with the recovery-ID of the registration -> ddnet.org verifies that it's your account -> ddnet.org makes new key-pair (either tells the game server about the change, or simply logs the change using the recovery-id so the game-server can ask if the identity changed from this to that) ( maybe: & generates a new recovery-key ) -> player can play again without hacker. <-- the client will fetch the new recovery-key and key-pair once the game-server told the client that recovery worked
- user cannot directly loose the account of the game-server: He can only loose the ddnet.org account. The rest is managed by this.
- Game-server only needs to communicate with ddnet.org when account is lost, else never

Pros:
- Easier to understand than the key-pair list
- Relatively easy to implement
- As few interaction with ddnet.org as possible. So relatively attractive for other mods
- Basically decentralized (except the recovery of the account), would even work offline after registration (or without ddnet.org it actually works offline).

Cons:
- More complex than a simple login/password approach for servers to implement
- ddnet.org recovery wrapper is extra work

I hope the idea is clear :)

**Comment by edg-l** [comment]
I'm in favor of the "automatic minimal account creation" and "fill in further details later" approach. Which works well with pub-key auth (without pass phrase?). It also works well with "optional accounts" if we consider the unfilled automatic account as a no-account kinda.

I also like steam approach with account IDs, you can have any name, you are only governed by an ID. But i acknowledge it has some drawbacks i guess, you would rely on client verification (friend list?) to know at first glance someone is someone, without having to manually check the ID.

I also think it's extremely important to have a good API around accounts, and being able to set which information a player exposes to it. E.g some people might want to disable showing how many hours they play, or where they play, etc.

**Comment by Jupeyy** [comment]
I think we should try a non representative small vote in this issue for developers only. We have many ideas here and I'd like to see how the tendency generally looks like.
Sadly most active devs already have an own proposal and are probably biased , so not sure how well it works

Let's not focus on the question if accounts at all. All methods will to some degree allow postponed account creation

👍 kyle-bradly - centralized Account system, mentions implementation details
❤️ Def - centralized Account system. User gets private key for auth. As far as I see ddnets account server would tell if the pub key is valid
😄Learath - from discord it sounds like he wants a classic centralized Account system with steam integration for easier registering.. so I assume for ddnet only(?)
🎉 Heinrich - accounts are decentralized, but a list of keys bound to the user are account is hosted by ddnet which all game servers can download. Recovery is optionaly doable by setting an email. Account names like discord(?)
🚀 Jupeyy - decentralized or semi-centralized. Let the user choose if they want to use ddnet register service which also supports steam login or classic email login. Allow fully decentralized account where user has to make sure not to loose keys(can be used for new players that don't register over ddnet service directly to automatically have an account). Pub key is used to indentify. Key pairs per server group. Other game servers don't need to know about ddnet except for account recovery. For normal users this should feel like a centralized Account system. Can still be implemented even when starting with centralized accounts

I guess for all system possible UI would be integrated into the client direclty

React with emojis. I tried to only use positive emojis

**Comment by Learath2** [comment]
FWIW, I'm not a huge fan of classical centralized authentication systems. I think I want something between what @Jupeyy proposed and what @heinrich5991 and I discussed a year ago.

Recovery is important, and users can't be trusted to back up keys properly.

On Discord I've been talking about a centralized one because it feels the most realistic, but ideally I also want shiny asymmetric crypto auth.

**Comment by Jupeyy** [comment]
> FWIW, I'm not a huge fan of classical centralized authentication systems. I think I want something between what @Jupeyy proposed and what @heinrich5991 and I discussed a year ago.


If u want simply edit my msg. I don't know perfectly what we all proposed in detail xD. There were so many with small detail differences

**Comment by ewancg** [comment]
> I think we should try a non representative small vote in this issue for developers only. We have many ideas here and I'd like to see how the tendency generally looks like. Sadly most active devs already have an own proposal and are probably biased , so not sure how well it works
> 
> Let's not focus on the question if accounts at all. All methods will to some degree allow postponed account creation
> 
> 👍 kyle-bradly - centralized Account system, mentions implementation details ❤️ Def - centralized Account system. User gets private key for auth. As far as I see ddnets account server would tell if the pub key is valid 😄Learath - from discord it sounds like he wants a classic centralized Account system with steam integration for easier registering.. so I assume for ddnet only(?) 🎉 Heinrich - accounts are decentralized(or partially? or are they at all?), but a list of keys bound to the user are account is hosted by ddnet which all game servers can download. Recovery is optionaly doable by setting an email. Account names like discord(?) 🚀 Jupeyy - decentralized or semi-centralized. Let the user choose if they want to use ddnet register service which also supports steam login or classic email login. Allow fully decentralized account where user has to make sure not to loose keys(can be used for new players that don't register over ddnet service directly to automatically have an account). Pub key is used to indentify. Key pairs per server group. Other game servers don't need to know about ddnet except for account recovery. For normal users this should feel like a centralized Account system. Can still be implemented even when starting with centralized accounts
> 
> I guess for all system possible UI would be integrated into the client direclty
> 
> React with emojis. I tried to only use positive emojis

Why developers only? Seems like a bad idea

**Comment by Jupeyy** [comment]
Well some developer has to add it anyway, also it's not meant as a strong vote, but i don't see feedback in this issue like "Omg this is so amazing lets do it".. So really just to see a tendency

**Comment by ChillerDragon** [comment]
> > React with emojis. I tried to only use positive emojis
> 
> 
> 
> Why developers only? Seems like a bad idea

It's not like non developers are technically excluded. Any non technical person who cares about the future of ddnet or accounts in particular can follow along on GitHub or make an account to react with an emoji.

**Comment by ewancg** [comment]
> > > React with emojis. I tried to only use positive emojis
> > 
> > 
> > Why developers only? Seems like a bad idea
> 
> It's not like non developers are technically excluded. Any non technical person who cares about the future of ddnet or accounts in particular can follow along on GitHub or make an account to react with an emoji.

That's true. It was mostly the wording that rubbed me the wrong way :smiley:

**Comment by Jupeyy** [comment]
I'd like to emphasize that I'd be willing to implement my idea after my holidays under the premise that it's not a wasted effort.
So if _you_ don't have an opinion yet, but want to build one, it would be nice if we could settle this debate during the next 2-3 weeks.

**Comment by heinrich5991** [comment]
How do you want to implement authentication against a server? It seems pretty impossible to me to do it in a way that doesn't allow for MITM without a secure server connection?

**Comment by Jupeyy** [comment]
i'll use the master server as trusted entity

**Comment by heinrich5991** [comment]
You have a trusted entity. How does the game server relate the client to an account?

**Comment by Jupeyy** [comment]
I already wrote about that

**Comment by heinrich5991** [comment]
> I already wrote about that

Probably referring to https://github.com/ddnet/ddnet/issues/3411#issuecomment-1312744483.

This design would imply that the client can only authenticate against servers if they're registered against the masterserver. If we'd allow to connect to a server by link, the auth information could not be verified and the account could be MITMed. This means that LAN servers and servers joined by link cannot be authenticated against from the client.

**Comment by Jupeyy** [comment]
> i'll use the master server as trusted entity

^

**Comment by ewancg** [comment]
If that is the case, servers connected to by link should just be expected not to use the accounts system. Third party game servers implementing the auth. will almost always be on the masters, and a MITM due to lack of secure connection isn't the only security concern presented by attempting to auth. on unverified servers.

**Comment by Jupeyy** [comment]
I definitely have concerns about what happens if a server gets compromised. E.g. how much damage can you deal. Not only if the game servers but also if ddnet gets hacked. I'd be open to start small and do it only for trusted servers first, where we can proof the identify unrelated to master server. That the connection must be secure always is clear to me tho. Also our community really isn't that big.

@heinrich5991 the 2-3 weeks I mentioned are really so we get to start on this. We drag this account stuff out to infinity. We can discuss out concerns, but we should also discuss what happens if we don't come further with concerns. E.g. start with normal centralized accounts, or drag the issue further into future.

I mean I am not even the biggest fan of accounts as you probably already noticed. But I can understand ppl that want this, and it doesn't seem like any dev is actually willing to go ahead and implement it finally^^

I proposed a solution that can probably build on centralized accounts too, if we setup everything it needs

**Comment by kaitlynia** [comment]
> Yeah, go and make accounts mandatory and see your favorite game dying because zero new players will come as the game is already complicated enough.

I understand that this is an old comment and much more productive things have been said since then, but I wanted to point out an issue I have with this common myth that is circulated among account-haters. New players do not expect the network to function the way it currently works. I have played with many new players that get confused and even frustrated after they change their name and it suddenly appears that they no longer have any points. What typically follows is me explaining that their points are bound to the name itself, and that really anyone could use that name and get points, and then they are seemingly even more frustrated after learning that those points are never really yours. Someone could fake you, or give you a finish for a map you really wanted to finish for yourself, or you could want a specific name that hasn't been used in years but you feel like it's not "fair" to use it if it already has points. All of these issues would be solved with accounts and it's minimally invasive the way it has been outlined in this thread numerous times... People exaggerate the effects of new features on the game's players, and despite so many changes to the game since the release on Steam you can clearly see that the stats on the website do not reflect "your favorite game" dying in any capacity.

**Comment by heinrich5991** [comment]
@Jupeyy and I talked about accounts a little today. He has implemented most of an account system and integrated it into ddnet-pg. It could also be made to work with our current DDNet, but it doesn't use a secure connection with the game server, given that we currently lack one.

He was thinking about porting it to DDNet, and we'd have to decide whether we'd want that. Would someone be interested in joining a discussion about that? @Learath2 maybe?

**Comment by heinrich5991** [comment]
To me, it feels like QUIC isn't that far away anymore, and I'd like to maybe wait, say, 3 more months to integrate the account system with TLS, so that it's safe from all™ attacks

**Comment by Learath2** [comment]
I would, but shouldn't he just outline how his account system works in a comment here so everyone gets to see it? Or did he just implement what he proposed here last year?

**Comment by Jupeyy** [comment]
> I would, but shouldn't he just outline how his account system works in a comment here so everyone gets to see it? Or did he just implement what he proposed here last year?

It's indeed mostly that +- some things i didn't like.

With the TLS handshake @heinrich5991 mentioned the concept would be slimmer, but still kinda similar (from database interaction on the game server POV).
It would require less talking to the account server and it would not require to build a key-pair per game server group(kog, ddnet etc.). It has some trades that aren't easy to explain in a few words, but the main idea is still that the login information comes directly from the client (by a certificate), in my current version it would directly be the public key (since no TLS).


So the question is mostly if we want to wait for TLS to have it slimmer and at the same time also more secure, but have to wait for the QUIC backend to be there (and kinda stable), or if we want to rush and finally get some account system done xdd

**Comment by Jupeyy** [comment]
Btw i'd also implement the other solution, I think it's generally better (even tho not all parts of it are clear yet), however what i'll not do then is:
- provide a FFI version for it (because i cannot clearly imagine how the interaction will look like)
- if some implementation details are not _good_ enough I don't want to promise that I'll be the person to fix them, since I cannot tell if I am still motivated for accounts in 3 months..

The current solution was mostly to try out and I decided to do it rather out of nowhere^^
So @Learath2 if you are really interested into understanding it, we could maybe talk it out in voice, and if others are interested too, we can also do it somewhere in a public discord voice chat or smth.
And i'd like to do it as quick as possible, not wait too long
